### ISSUE TITLE
[css-cascade-6] Add ability to scope rules from an imported stylesheet

### ISSUE THREAD

**Issue by faceless2** [w3c/csswg-drafts#7348]
This is following on from a brief conversation with @tabatkins at CSSDay.

While I like the scoping proposal, I think it really needs a way to scope styles from an imported stylesheet. As it is now you can define your scopes - great for modularization - but the rules have to be inline in the same sheet, which seems to defeat the purpose a bit to me.

There are several ways to do this - an `@import` inside a `@scope` is one, but I think for consistency with cascade layers the syntax should be like

```css
@import uri(module.css) scope(.media-object) to (.content);
```

Also useful but slightly more controversial (it doesn't have the desired behaviour when the attribute is not recognised) would be adding a `scope` and (pending a better idea) a `scope-to` attribute to `<link>`, the same way that it now has `layer` (see https://github.com/whatwg/html/pull/7658)

```html
<link rel="stylesheet" href="module.css" scope=".media-object" scope-to=".content">
```

**Comment by tabatkins** [comment]
Yeah, this seems reasonable to me. Might quibble a bit on the syntax, but being able to apply a scope to the rules in a stylesheet seems at least as useful as being able to put a stylesheet on a particular cascade layer.

**Comment by DarkWiiPlayer** [comment]
I think an easier way of handling scope in HTML would simply be:

```html
<link rel="stylesheet" href="module.css" scope=".media-object to .content">
```

**Comment by mirisuzanne** [comment]
My proposal would be that both a `scope()` function in CSS `@import`, and a `scope` attribute on the `<link>` element could use accept the full `@scope` condition syntax:

```css
@import uri(module.css) scope((.media-object) to (.content) inclusive);
```
```html
<link rel="stylesheet" href="module.css" scope="(.media-object) to (.content) exclusive">
```

The only edge case that might warrant a special-case is importing a scope with _only the scope root_. Without any exceptions, that would result in double-parentheses:

```css
@import uri(module.css) scope((.media-object));
```

---

I also want to recognize that adding attributes like this to HTML `<link>` also requires our proposed (but not yet fully specified or approved) update to the `<link media="">` attribute, so that it's possible to test for support of scoping, and only load the linked CSS file _if the scoping will be applied_. This is also needed for a proposed `layer` attribute.

There's a [PR in progress](https://github.com/whatwg/html/pull/7658/), but I could use help on some of the details.

**Comment by bramus** [comment]
@mirisuzanne I see that you removed the https://github.com/w3c/csswg-drafts/labels/Needs%20Edits label a while ago, but there does not seem to be a commit that added text to the spec for this resolution. Could it be this label was removed by mistake?

**Comment by mirisuzanne** [comment]
@bramus yes that's possible, and seems likely - it accidentally got associated with implicit scope issue.

**Comment by romainmenke** [comment]
Where would the `scope()` function be in the syntax of `@import` relative to `supports()` ?
I am assuming it will not be placed before `layer`, so either before or after `supports()`?

**Comment by bramus** [comment]
I don’t think the order in which these are declared should matter. E.g. `@import url(…) supports(…) scope(…)` and `@import url(…) scope(…) supports(…)` should behave the same, no?

The order in which they are processed could be fixed though, similar to how the individual transform properties are applied in a predetermined order.

**Comment by romainmenke** [comment]
Currently the order does matter for the existing parts :
1. `layer`
2. `supports()`
3. media queries list

Any other order is invalid.

----

I agree that a non-fixed order might be better.
Makes it easier for authors to get it right.

> The order in which they are processed could be fixed though, similar to how the individual transform properties are applied in a predetermined order.

Probably safe to specify this, but I don't think it can have observable side-effects if different browsers apply conditions in different orders. (`layer` must still be applied at a specific point relative to all the conditions)

The order also isn't specified today for `supports()` and media query lists : https://drafts.csswg.org/css-cascade-5/#conditional-import

Unless I am overlooking it?

**Comment by mirisuzanne** [comment]
Do we need to put this back on the agenda to discuss the questions about ordering?

**Comment by romainmenke** [comment]
Can we make it so that authors do not have to wrap simple scope conditions?

```css
@imports "foo" @scope(.foo); /* -> authors will want to write this */
@imports "foo" @scope((.foo)); /* explicitly wrapped in parenthesis */
@imports "foo" @scope((.foo) to (.bar));
```

**Comment by romainmenke** [comment]
`allow full reordering of all the current conditions on @import (MQ, scope, ...)`

There is an issue with allowing any order.
Media query lists can be partially valid.

Any media query list with more than one media query would swallow unrecognized trailing parts.

```css
@import url("foo") screen, print something-invalid(really-does-not-exist);
```

This still applies fine everywhere that matches `screen`.
The same for:

```css
@import url("foo") screen, print scope(.foo);
```

You can see this in action in this codepen:

https://codepen.io/romainmenke/pen/WNVvyBY

-----

Can we instead just allow any order between new conditions?
- `layer` always first
- media query lists always last
- anything in between can go in any order

**Comment by DarkWiiPlayer** [comment]
If I remember correctly, `@scope { /* styles */ }` to mean "scope to the parent of the style tag", would `<link scope src="whatever">` work analogously then? That seems like something that could be convenient in practice.

**Comment by mirisuzanne** [comment]
@DarkWiiPlayer I agree that could be useful. We didn't clearly specify how it should work here, so probably should work through those details.

@romainmenke oh that's interesting.

**Comment by ByteEater-pl** [comment]
@DarkWiiPlayer @fantasai

> I think an easier way of handling scope in HTML would simply be:
> 
> <link rel="stylesheet" href="module.css" scope=".media-object to .content">

It doesn't mean what you seem to think, but rather elements with class `content` having an ancestor of type `to` having an ancestor with class `media-object`.

**Comment by mirisuzanne** [comment]
That linked comment from @kizu raises a new issue that I'm not sure we've discussed: what happens with name-defining at-rules in a scoped import?

**Comment by kizu** [comment]
Note: right now, browsers just allow defining name-defining at-rules inside the `@scope`, which work as if they were defined outside — https://codepen.io/kizu/pen/jEOVrEN

So, following the same logic, nothing should change and these will work as if they were defined outside of the scope.

**Comment by faceless2** [comment]
>  The only other "reasonable" behaviour I can think of on the spot is that the name would only be valid inside that scope or override any definition from an outer scope, but that also seems like it could get incredibly confusing really quickly.

Surely that's the author's intent though? If a `@font-face` is defined in a scope, it's presumably because they only intended it to apply within that scope - when the scope ends, the rule no longer applies? I do realise this makes it a much harder problem.
